### ISSUE TITLE
Alternative nufft

### ISSUE THREAD

**Issue by garrettwrong** [ComputationalCryoEM/ASPIRE-Python#276]
It would be nice to have a pure python, or at least a non FI nufft implementation, in and regularly tested with the code, regardless of speed.

There was pynufft and another option as well.  It has also been suggested we can just compute things directly.  This is mainly just a consideration for continuity and reference purposes.

Initially thought we'd get there as part of cleanup, but I think we might want to punt on this short term. Converted from note to issue.

**Comment by garrettwrong** [comment]
Lifting this issue after the recent bug in FINUFFT.  I think we should do the following:

i) confirm/deny the state of the other (CPU) nufft implementations in our config. Last I checked only FINUFFT worked.
ii) find and maintain one working nufft alternative in ASPIRE
iii) implement a nuft (the slow explicit one) as a reference/fall back. This is practically useless but can come in handy for debug/reference. 

iii) This was the original intent of this issue, but I think it is reasonable to expand on it.

**Comment by janden** [comment]
I agree with (i) and (iii), but (ii) can be hard. The issue is that the time is dominated by unstructured memory access, which means that a naive NUFFT implementation in Python is likely to be _slower_ than a “slow” NUDFT (i.e., option (iii)). If we really want to do it, we can take some inspiration from https://github.com/jakevdp/nufftpy. It's still going to be a lot of work.

**Comment by garrettwrong** [comment]
https://github.com/pyNFFT/pyNFFT does not look active, so I don't know if it is worth necromancing it in ASPIRE.  Probably should remove instead.

It looks like https://github.com/jyhmiinlin/pynufft is active, so we might be able to use that...

I was also thinking that I could do a small project with the guys that just implements the nudft for 2d and 3d in a naive way using CUDA.   If we just went right for type 3 and didn't concern with performance too much, might be reasonably scoped.  This would take some time, with a goal to train them more on GPU, but we would get some useful software out of it...  For our small (unit test sized) problems the asymptotics should be somewhat offset by CUDA cores.  I could see other people using it for comparison if the accuracy/correctness is really tight...

In any case, I think with some of the major movements planned in the Flatiron codes we should probably have an alternative ready.

**Comment by garrettwrong** [comment]
Noting that I attempted quickly adding `pynufft`, but the gpu implementation in package appears to be behind the times (relies on openCL underneath), and so that one probably will not help us.
